sunpy/coordinates/spice.py

```diff
@@ -0,0 +1,217 @@
+
+The `SPICE <https://naif.jpl.nasa.gov/naif/>`_ observation geometry
+information system is being increasingly used by space missions to describe the
+locations of spacecraft and the time-varying orientations of reference frames.
+While the `~spiceypy.spiceypy` package provides a Python interface for
+performing SPICE computations, its API is very different from that of
+`~astropy.coordinates.SkyCoord`.
+
+This module "wraps" `~spiceypy.spiceypy` functionality so that relevant SPICE
+computations can be accessed using the `~astropy.coordinates.SkyCoord` API.
+When loading a set of kernels, a frame class and corresponding transformations
+are created for each SPICE frame.  One can also query the location of a body
+as computed via SPICE.
+
+.. note::
+    This module requires the optional dependency `~spiceypy.spiceypy` to be
+    installed.
+
+Notes
+-----
+* All transformations from one SPICE frame to another SPICE frame go through
+  `~astropy.coordinates.ICRS` as the intermediate frame, even if the origin
+  shift to/from the solar-system barycenter is unnatural.
+* Transformations of velocities are not yet supported.
+* SPICE frame names are rendered as uppercase, except for plus/minus characters,
+  which are replaced with lowercase ``'p'``/``'n'`` characters.
+"""
+
+try:
+    import spiceypy
+except ImportError:
+    raise ImportError("This module requires the optional dependency `spiceypy`.")
+
+import astropy.units as u
+from astropy.coordinates import ICRS, SkyCoord, frame_transform_graph
+from astropy.coordinates.representation import CartesianRepresentation
+from astropy.coordinates.transformations import FunctionTransformWithFiniteDifference
+from astropy.time import Time
+
+from sunpy import log
+from sunpy.coordinates import SunPyBaseCoordinateFrame
+from sunpy.time import parse_time
+from sunpy.time.time import _variables_for_parse_time_docstring
+from sunpy.util.decorators import add_common_docstring
+
+__all__ = ['get_body', 'initialize']
+
+
+# Note that this epoch is very slightly different from the typical definition of J2000.0 (in TT)
+_ET_REF_EPOCH = Time('J2000', scale='tdb')
+
+_CLASS_TYPES = {2: 'PCK', 3: 'CK', 4: 'TK', 5: 'dynamic', 6: 'switch'}
+
+
+# Registry of the generated frame classes
+spice_frame_classes = []
+
+
+# Defined for future use
+class _SpiceBaseCoordinateFrame(SunPyBaseCoordinateFrame):
+    pass
+
+
+def _convert_to_et(time):
+    return (time - _ET_REF_EPOCH).to_value('s')
+
+
+def _install_frame_by_id(frame_id):
+    frame_name = spiceypy.frmnam(frame_id)
+
+    # Replace plus/minus characters in the SPICE frame name with lowercase 'p'/'n'
+    frame_name = frame_name.translate(str.maketrans("+-", "pn"))
+
+    frame_center, class_num, _ = spiceypy.frinfo(frame_id)
+    frame_center_name = spiceypy.bodc2n(frame_center)
+    log.info(f"Installing {frame_name} {_CLASS_TYPES[class_num]} frame ({frame_id}) "
+             f"as 'spice_{frame_name}'")
```

Comment:
Does any of this need some error handling? Can you end up in a situation where you request a frame_id which doesn't exist?

Reply:
As a private function, it doesn't require any error handling, because the calling code will be handing it valid frame IDs.  And besides, if it somehow receives an invalid frame ID, `spicepy` will raise an error at the first line with `frmnam()`, so that's your validity check right there.